update services guide overview (#993)
crossplaneio/crossplane: docs/services-guide.md

classes can reference secure connectivity resources (networks), such that new
instances of that class can be made available on the cluster's private network.
Resource classes, cluster networking, and secure connectivity resources are
designed to work together to enable self-service provisioning of securely
connected cloud services in a Kubernetes cluster.

Resource claims can be matched to a class in several ways: 
1. rely on a class marked  `resourceclass.crossplane.io/is-default-class:
"true"`
1. match on class labels using a `claim.spec.classSelector` 
1. use a `claim.spec.classRef` to a specific class

The first two methods rely on a default class of service or use a
`classSelector` that matches any suitable resource class available in the
target cluster.  As such, the first two methods are considered portable
resource claims that can be used in any cluster that provides the desired class
of service. You may have one cluster using GCP and another cluster using AWS,
and the same claim can be used in either cluster so long as the claim can be
matched to a suitable class of cloud service.

The third method uses an explicit `classRef` to a specific resource class like

> `CloudSQLInstanceClass` which means the claim may only be used with that class.

Reviewer: ```suggestion
a `CloudSQLInstanceClass` which means the claim may only be used with that class.
```

weight: 410
---
# Services Guide
This guide is an overview of enabling cloud service 
provisioning on an existing Kubernetes target cluster, 

> including how to integrate Crossplane with existing cloud networking configurations 

Reviewer: nitpick: word-wrap looks like it might be acting strangely on these lines here. I could be wrong though

to provide secure managed service connectivity.  Step-by-step instructions are 
provided for [GCP][gcp-services-guide], [AWS][aws-services-guide], 
and [Azure][azure-services-guide].

To dynamically provision a new Kubernetes target cluster 
see the Stacks Guides for [GCP][stack-guide-gcp], 
[AWS][stack-guide-aws], and [Azure][stack-guide-azure].

## Table of Contents
1. [Introduction](#introduction)
1. [Secure network connectivity for cloud services](#secure-network-connectivity-for-cloud-services)
1. [Dynamic provisioning with claims and classes](#dynamic-provisioning-with-claims-and-classes)
1. [Connection secret for pods in a deployment](#connection-secret-for-pods-in-a-deployment)
1. [Next Steps](#next-steps)
1. [Learn More](#learn-more)

## Introduction
Cloud service provisioning can be added to existing clusters by installing 
Crossplane directly onto the target cluster. Crossplane is designed to 
integrate with existing cloud networking and security resources, so 
managed services like RDS, CloudSQL, and Azure DB can be provisioned 

> using Kubernetes objects securly consumed by pods in a cluster. 

Reviewer: ```suggestion
using Kubernetes objects and secuerly consumed by pods in a cluster. 
```

Crossplane achieves this by:
1. establishing secure network connectivity between the worker nodes in a cluster
and the cloud services using a set of networking and security resources
1. populating Kuberentes `Secrets` that pods in a `Deployment` can use to 
securely access the managed service

## Secure network connectivity for cloud services
Crossplane currently supports private IP secure connectivity for
AWS, GCP, and Azure Stacks. Managed services instances

> are made available on the cluster's prviate network(s) so pods can access them.

Reviewer: nitpick: the word-wrap looks strange to me here as well

Crossplane also supports configuring ingress/egress rules to further restrict 
allowed network traffic.

While each cloud provider uses different resources for establishing 
secure connectivity between a Kubernetes cluster (EKS, GKE, AKS) 
and managed services (RDS, CloudSQL, and Azure DB), the basic pattern is the
same:
1. Configure cluster networking
   * network(s) and subnet(s) - L3 networking for the worker nodes
1. Enable managed service access: 
   * private service connection / endpoint - make services available via peering or other
   * private IP range(s) or subnet group - the private IPs a managed service will get
   * security groups or network rules - to restrict network traffic
1. Provision a managed service instance
   * creates an instance (e.g. RDS, CloudSQL, Azure DB MySQL instance)
   * assigns a private IP from the private IP range above
1. Securely use the managed service with secrets
   * pods on a cluster node can access the managed service via private IP

> Crossplane provides Kubernetes objects for all of the above, so you can

Reviewer 2: We tend to say "Kubernetes resources" rather than "objects" in most places. It's hard to determine which is more correct; for example https://blog.openshift.com/kubernetes-deep-dive-api-server-part-1/ and https://stackoverflow.com/a/52314551 seem to contradict each other.

define a secure connectivity model for the managed services you want to make 
available for self-service provisioning in the cluster using claims and classes.

## Dynamic provisioning with claims and classes
Crossplane employs a layered architecture consisting of managed resources that
represent a cloud service, and resource claims and classes that enable dynamic

> provisioning on top.

Reviewer: Maybe something like "enable portable dynamic provisioning of those services".

Reviewer 2: When writing the getting started guide I felt like we use a lot of terminology that may not make sense to newcomers, and wondered whether we should avoid that terminology in overviews, headings, etc. Terms like "claim", "dynamic provisioning" and even perhaps "class" don't seem intuitive to me. I tried to address this by either avoiding those terms in introductory contexts or trying to use them in contexts that might suggest what they mean.

Managed resources are high fidelity representations of the API resources that 
make up a cloud service. They're not portable across clouds. 
A `CloudSQLInstance` is an example of a managed resource - it's relevant only 
to the Google Cloud Platform (GCP) and exposes all of the 
nitty gritty configuration details of a CloudSQL instance. The networking
and security Kubernetes objects mentioned above fall into this category.

Resource claims and classes are the next layer up. Resource claims like 
`MySQLInstance` enable dynamic provisioning of managed resources by matching 
a claim to a class like `CloudSQLInstanceClass` that provides the detailed configuration 
template to provisoin a new cloud service instance. Resource classes can reference
secure connectivity resources (networks), such that new instances of 

> that class can be made available on the clusters private network. Resource classes,

Reviewer: ```suggestion
that class can be made available on the cluster's private network. Resource classes,
```

cluster networking, and secure connectivity resources are designed to work together
to enable self-service provisoining of securely connected cloud services in a 
Kubernetes cluster.

Resource claims can be matched to a class in several ways: 
1. rely on a class marked  `resourceclass.crossplane.io/is-default-class: "true"`
1. match on class labels using a `claim.spec.classSelector` 
1. use a `claim.spec.classRef` reference a specific class

The first two methods rely on a default class of service or use a `classSelector`
that matches any suitable resource class available in the target cluster. 
As such, the first two methods are considered portable resource claims that
can be used in any cluster that provides the desired class of service. You may have
one cluster using GCP and another cluster using AWS, and the same claim can be used
in either cluster so long as the claim can be matched to a suitable class of cloud service.

> The thirid method uses an explicit `classRef` to a specific resource class

Reviewer: ```suggestion
The third method uses an explicit `classRef` to a specific resource class
```

> like `CloudSQLInstanceClass` which means the claim may only be used with that class.

Reviewer: ```suggestion
like a `CloudSQLInstanceClass` which means the claim may only be used with that class.
```
